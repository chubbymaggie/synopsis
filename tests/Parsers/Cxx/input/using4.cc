// From C++WD'96 7.4.3.6 Example
namespace D
{
  int d1;
  void f(char);
}

using namespace D;
int d1; // ok, no conflict with D::d1
namespace E
{
  int e;
  void f(int);
}

namespace D  // namespace extension
{
  int d2;
  using namespace E;
  void f(int);
}

void f()
{
  d1++;    // error: ambiguous: ::d1 or D::d1
  ::d1++;  // ok
  D::d1++; // ok
  d2++;    // ok D::d2
  e++;     // ok E::e
  f(1);    // error: ambiguous: D::f(int) or E::f(int)
  f('a');  // ok: D::f(char)
}
