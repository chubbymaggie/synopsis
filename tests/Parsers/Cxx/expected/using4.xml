<?xml version='1.0' encoding='ISO-8859-1'?>
<?xml-stylesheet href='/home/stefan/synopsis/Synopsis/share/Synopsis/dump.css' type='text/css'?>
<ast>
 <declarations>
  <instance accessibility="0" class="Synopsis.AST.Module" file="Parsers/Cxx/input/using4.cc" language="C++" line="0" name="D" type="namespace">
   <comments>
    <instance class="Synopsis.AST.Comment" line="0" suspect="0">
     <file>
      Parsers/Cxx/input/using4.cc
     </file>
     <text>
      // From C++WD'96 7.4.3.6 Example

     </text>
    </instance>
   </comments>
   <declarations>
    <instance accessibility="0" class="Synopsis.AST.Variable" constr="0" file="Parsers/Cxx/input/using4.cc" language="C++" line="4" name="D.d1" type="namespace variable">
     <vtype>
      <instance class="Synopsis.Type.Base" language="C++" name="int"/>
     </vtype>
    </instance>
    <instance accessibility="0" class="Synopsis.AST.Function" file="Parsers/Cxx/input/using4.cc" language="C++" line="5" name="D.f(char)" realname="f" type="function">
     <parameters>
      <instance class="Synopsis.AST.Parameter" identifier="" value="">
       <type>
        <instance class="Synopsis.Type.Base" language="C++" name="char"/>
       </type>
      </instance>
     </parameters>
     <returnType>
      <instance class="Synopsis.Type.Base" language="C++" name="void"/>
     </returnType>
    </instance>
    <instance accessibility="0" class="Synopsis.AST.Declaration" file="Parsers/Cxx/input/using4.cc" language="C++" line="5" name="dummy" type="dummy"/>
    <instance accessibility="0" class="Synopsis.AST.Variable" constr="0" file="Parsers/Cxx/input/using4.cc" language="C++" line="18" name="D.d2" type="namespace variable">
     <comments>
      <instance class="Synopsis.AST.Comment" line="0" suspect="0">
       <file>
        Parsers/Cxx/input/using4.cc
       </file>
       <text>
        // namespace extension

       </text>
      </instance>
     </comments>
     <vtype/>
    </instance>
    <instance accessibility="0" class="Synopsis.AST.Function" file="Parsers/Cxx/input/using4.cc" language="C++" line="20" name="D.f(int)" realname="f" type="function">
     <parameters>
      <instance class="Synopsis.AST.Parameter" identifier="" value="">
       <type/>
      </instance>
     </parameters>
     <returnType/>
    </instance>
    <instance accessibility="0" class="Synopsis.AST.Declaration" file="Parsers/Cxx/input/using4.cc" language="C++" line="20" name="dummy" type="dummy"/>
   </declarations>
  </instance>
  <instance accessibility="0" class="Synopsis.AST.Variable" constr="0" file="Parsers/Cxx/input/using4.cc" language="C++" line="9" name="d1" type="global variable">
   <vtype/>
  </instance>
  <instance accessibility="0" class="Synopsis.AST.Module" file="Parsers/Cxx/input/using4.cc" language="C++" line="0" name="E" type="namespace">
   <comments>
    <instance class="Synopsis.AST.Comment" line="0" suspect="0">
     <file>
      Parsers/Cxx/input/using4.cc
     </file>
     <text>
      // ok, no conflict with D::d1

     </text>
    </instance>
   </comments>
   <declarations>
    <instance accessibility="0" class="Synopsis.AST.Variable" constr="0" file="Parsers/Cxx/input/using4.cc" language="C++" line="12" name="E.e" type="namespace variable">
     <vtype/>
    </instance>
    <instance accessibility="0" class="Synopsis.AST.Function" file="Parsers/Cxx/input/using4.cc" language="C++" line="13" name="E.f(int)" realname="f" type="function">
     <parameters>
      <instance class="Synopsis.AST.Parameter" identifier="" value="">
       <type/>
      </instance>
     </parameters>
     <returnType/>
    </instance>
    <instance accessibility="0" class="Synopsis.AST.Declaration" file="Parsers/Cxx/input/using4.cc" language="C++" line="13" name="dummy" type="dummy"/>
   </declarations>
  </instance>
  <instance accessibility="0" class="Synopsis.AST.Function" file="Parsers/Cxx/input/using4.cc" language="C++" line="23" name="f()" realname="f" type="function">
   <returnType/>
  </instance>
 </declarations>
 <types>
  <instance class="Synopsis.Type.Declared" language="C++" name="E.e">
   <declaration/>
  </instance>
  <instance class="Synopsis.Type.Declared" language="C++" name="D.f(int)">
   <declaration/>
  </instance>
  <instance class="Synopsis.Type.Declared" language="C++" name="E.f(int)">
   <declaration/>
  </instance>
  <instance class="Synopsis.Type.Declared" language="C++" name="dummy">
   <declaration/>
  </instance>
  <instance class="Synopsis.Type.Declared" language="C++" name="D.f(char)">
   <declaration/>
  </instance>
  <instance class="Synopsis.Type.Declared" language="C++" name="D">
   <declaration/>
  </instance>
  <instance class="Synopsis.Type.Declared" language="C++" name="E">
   <declaration/>
  </instance>
  <instance class="Synopsis.Type.Declared" language="C++" name="f()">
   <declaration/>
  </instance>
  <instance class="Synopsis.Type.Declared" language="C++" name="d1">
   <declaration/>
  </instance>
  <instance class="Synopsis.Type.Declared" language="C++" name="D.d2">
   <declaration/>
  </instance>
  <instance class="Synopsis.Type.Declared" language="C++" name="D.d1">
   <declaration/>
  </instance>
 </types>
 <files>
  <file>
   Parsers/Cxx/input/using4.cc
  </file>
 </files>
</ast>
